attributecert/attributecert.go

```diff
@@ -0,0 +1,712 @@
+	if err != nil {
+		return nil, err
+	}
+
+	var issuer pkix.RDNSequence
+	if rest, err := asn1.Unmarshal(v.Bytes, &issuer); err != nil {
+		return nil, err
+	} else if len(rest) != 0 {
+		return nil, errors.New("attributecert: trailing data after X.509 subject")
+	}
+
+	out.Issuer.FillFromRDNSequence(&issuer)
+	if _, err := asn1.Unmarshal(in.TBSAttributeCertificate.Holder.BaseCertificateID.Issuer.Bytes, &v); err != nil {
+		return nil, err
+	}
+
+	var holder pkix.RDNSequence
+	if rest, err := asn1.Unmarshal(v.Bytes, &holder); err != nil {
+		return nil, err
+	} else if len(rest) != 0 {
+		return nil, errors.New("attributecert: trailing data after X.509 subject")
+	}
+
+	out.Holder.Issuer.FillFromRDNSequence(&holder)
+	out.Holder.Serial = in.TBSAttributeCertificate.Holder.BaseCertificateID.Serial
+
+	out.NotBefore = in.TBSAttributeCertificate.Validity.NotBefore
+	out.NotAfter = in.TBSAttributeCertificate.Validity.NotAfter
+
+	for _, attribute := range in.TBSAttributeCertificate.Attributes {
+		switch {
+		case attribute.Id.Equal(oidAttributeUserNotice):
+			_, err := asn1.Unmarshal(attribute.RawValues[0].FullBytes, &out.UserNotice)
+			if err != nil {
+				return nil, err
+			}
+		case attribute.Id.Equal(oidTcgPlatformSpecification):
+			_, err := asn1.Unmarshal(attribute.RawValues[0].FullBytes, &out.TCGPlatformSpecification)
+			if err != nil {
+				return nil, err
+			}
+		case attribute.Id.Equal(oidTbbSecurityAssertions):
+			_, err := asn1.Unmarshal(attribute.RawValues[0].FullBytes, &out.TBBSecurityAssertions)
+			if err != nil {
+				return nil, err
+			}
+		case attribute.Id.Equal(oidTcgCredentialSpecification):
+			var credentialSpecification TCGCredentialSpecification
+			_, err := asn1.Unmarshal(attribute.RawValues[0].FullBytes, &credentialSpecification)
+			if err != nil {
+				return nil, err
+			}
+		case attribute.Id.Equal(oidTcgCredentialType):
+			var credentialType TCGCredentialType
+			_, err := asn1.Unmarshal(attribute.RawValues[0].FullBytes, &credentialType)
+			if err != nil {
+				return nil, err
+			}
+		case attribute.Id.Equal(oidTcgPlatformConfigurationV1):
+			var platformConfiguration PlatformConfigurationV1
+			_, err := asn1.Unmarshal(attribute.RawValues[0].FullBytes, &platformConfiguration)
+			if err != nil {
+				return nil, err
+			}
+		case attribute.Id.Equal(oidTcgPlatformConfigurationV2):
+			var platformConfiguration PlatformConfigurationV2
+			_, err := asn1.Unmarshal(attribute.RawValues[0].FullBytes, &platformConfiguration)
+			if err != nil {
+				return nil, err
+			}
+		case attribute.Id.Equal(oidTcgPlatformConfigUri):
+			var platformConfigurationUri UriReference
+			_, err := asn1.Unmarshal(attribute.RawValues[0].FullBytes, &platformConfigurationUri)
+			if err != nil {
+				return nil, err
+			}
+		default:
+			return nil, fmt.Errorf("attributecert: unknown attribute %v", attribute.Id)
+		}
+	}
+
+	for _, extension := range in.TBSAttributeCertificate.Extensions {
+		switch {
```

Comment:
Nit: Please add a newline between clauses of the outmost switch block, so its a little easier to visually scan through.